deploy/src/controller/cluster.rs

```diff
@@ -0,0 +1,369 @@
+use std::collections::BTreeMap;
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use k8s_openapi::api::apps::v1::{
+    RollingUpdateStatefulSetStrategy, StatefulSet, StatefulSetSpec, StatefulSetUpdateStrategy,
+};
+use k8s_openapi::api::batch::v1::{CronJob, CronJobSpec, JobSpec, JobTemplateSpec};
+use k8s_openapi::api::core::v1::{
+    Container, PersistentVolumeClaim, PodSpec, PodTemplateSpec, Service, ServicePort, ServiceSpec,
+    VolumeMount,
+};
+use k8s_openapi::apimachinery::pkg::apis::meta::v1::{LabelSelector, ObjectMeta, OwnerReference};
+use k8s_openapi::apimachinery::pkg::util::intstr::IntOrString;
+use kube::api::{Patch, PatchParams};
+use kube::{Api, Client, Resource};
+use tracing::{debug, error};
+
+use utils::consts::{DEFAULT_BACKUP_DIR, DEFAULT_DATA_DIR};
+
+use crate::controller::consts::FIELD_MANAGER;
+use crate::controller::Controller;
+use crate::crd::{Cluster, StorageSpec};
+
+/// CRD `XlineCluster` controller
+pub(crate) struct ClusterController {
+    /// Kubernetes client
+    pub(crate) kube_client: Client,
+    /// Cluster api
+    pub(crate) cluster_api: Api<Cluster>,
+    /// The kubernetes cluster dns suffix
+    pub(crate) cluster_suffix: String,
+}
+
+/// All possible errors
+#[derive(thiserror::Error, Debug)]
+pub(crate) enum Error {
+    /// Missing an object in cluster
+    #[error("Missing object key {0} in cluster")]
+    MissingObject(&'static str),
+    /// Kube error
+    #[error("Kubernetes api error")]
+    Kube(#[from] kube::Error),
+    /// Backup PV mount path is already mounted
+    #[error("The path {0} is internally used in the xline operator and cannot be mounted.")]
+    CannotMount(&'static str),
+}
+
+/// Controller result
+type Result<T> = std::result::Result<T, Error>;
+
+impl ClusterController {
+    /// Extract service ports
+    fn extract_service_ports(cluster: &Arc<Cluster>) -> Result<Vec<ServicePort>> {
+        // expose all the container pods
```

Comment:
Is it a typo? "expose all the container ports"

Reply:
done